beacon-chain/db/kv/blob.go

```diff
@@ -80,61 +75,95 @@ func (s *Store) SaveBlobSidecar(ctx context.Context, scs []*ethpb.BlobSidecar) e
-func (s *Store) verifySideCars(scs []*ethpb.BlobSidecar) error {
+var (
+	errBlobSlotMismatch     = errors.New("sidecar slot mismatch")
+	errBlobParentMismatch   = errors.New("sidecar parent root mismatch")
+	errBlobRootMismatch     = errors.New("sidecar root mismatch")
+	errBlobProposerMismatch = errors.New("sidecar proposer index mismatch")
+	errBlobSidecarLimit     = errors.New("sidecar exceeds maximum number of blobs")
+	errEmptySidecar         = errors.New("nil or empty blob sidecars")
+)
+
+// validUniqueSidecars ensures that all sidecars have the same slot, parent root, block root, and proposer index, and no more than MAX_BLOB_EPOCHS.
+func validUniqueSidecars(scs []*ethpb.BlobSidecar) ([]*ethpb.BlobSidecar, error) {
 	if len(scs) == 0 {
-		return errors.New("nil or empty blob sidecars")
-	}
-	if uint64(len(scs)) > fieldparams.MaxBlobsPerBlock {
-		return fmt.Errorf("too many sidecars: %d > %d", len(scs), fieldparams.MaxBlobsPerBlock)
+		return nil, errEmptySidecar
 	}
 
-	sl := scs[0].Slot
-	pr := scs[0].BlockParentRoot
-	r := scs[0].BlockRoot
-	p := scs[0].ProposerIndex
+	// If there's only 1 sidecar, we've got nothing to compare.
+	if len(scs) == 1 {
+		return scs, nil
+	}
 
-	for _, sc := range scs {
-		if sc.Slot != sl {
-			return fmt.Errorf("sidecar slot mismatch: %d != %d", sc.Slot, sl)
+	prev := scs[0]
+	didx := 1
+	for i := 1; i < len(scs); i++ {
+		sc := scs[i]
+		if sc.Slot != prev.Slot {
+			return nil, errors.Wrapf(errBlobSlotMismatch, "%d != %d", sc.Slot, prev.Slot)
+		}
+		if !bytes.Equal(sc.BlockParentRoot, prev.BlockParentRoot) {
+			return nil, errors.Wrapf(errBlobParentMismatch, "%x != %x", sc.BlockParentRoot, prev.BlockParentRoot)
+		}
+		if !bytes.Equal(sc.BlockRoot, prev.BlockRoot) {
+			return nil, errors.Wrapf(errBlobRootMismatch, "%x != %x", sc.BlockRoot, prev.BlockRoot)
 		}
-		if !bytes.Equal(sc.BlockParentRoot, pr) {
-			return fmt.Errorf("sidecar parent root mismatch: %x != %x", sc.BlockParentRoot, pr)
+		if sc.ProposerIndex != prev.ProposerIndex {
+			return nil, errors.Wrapf(errBlobProposerMismatch, "%d != %d", sc.ProposerIndex, prev.ProposerIndex)
 		}
-		if !bytes.Equal(sc.BlockRoot, r) {
-			return fmt.Errorf("sidecar root mismatch: %x != %x", sc.BlockRoot, r)
+		// skip duplicate
+		if sc.Index == prev.Index {
+			continue
 		}
-		if sc.ProposerIndex != p {
-			return fmt.Errorf("sidecar proposer index mismatch: %d != %d", sc.ProposerIndex, p)
+		if didx != i {
+			scs[didx] = scs[i]
 		}
+		prev = scs[i]
+		didx += 1
 	}
-	return nil
+
+	if didx > fieldparams.MaxBlobsPerBlock {
+		return nil, errors.Wrapf(errBlobSidecarLimit, "%d > %d", didx, fieldparams.MaxBlobsPerBlock)
+	}
+	return scs[0:didx], nil
 }
 
 // sortSideCars sorts the sidecars by their index.
-func sortSideCars(scs []*ethpb.BlobSidecar) {
-	sort.Slice(scs, func(i, j int) bool {
-		return scs[i].Index < scs[j].Index
+func sortSideCars(sc []*ethpb.BlobSidecar) {
```

Comment:
small comment, any reason for changing the name here? just shorter?

Reply:
oh wait, I had some changes to this func that I wound up reverting but this bit lingered. I meant to also change the capitalization of the func. But actually this is just unrelated cruft. 

The change I had for a bit was to take the BlobSidecars container. That changed the argument from a list of sidecars to a single sidecars value. So plural -> singular.

I'll go ahead and change the capitalization since we've already talked about it and undo the variable name change :)

Reply:
updated to revert that variable name change and change the func name, please take another look.